scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala

```diff
@@ -0,0 +1,427 @@
+      allowEnums: Boolean = this.allowEnums,
+      allowImplicitByNameParameters: Boolean = this.allowImplicitByNameParameters,
+      allowImplicitFunctionTypes: Boolean = this.allowImplicitFunctionTypes,
+      allowInlineIdents: Boolean = this.allowInlineIdents,
+      allowInlineMods: Boolean = this.allowInlineMods,
+      allowLiteralTypes: Boolean = this.allowLiteralTypes,
+      allowMethodTypes: Boolean = this.allowMethodTypes,
+      allowMultilinePrograms: Boolean = this.allowMultilinePrograms,
+      allowOrTypes: Boolean = this.allowOrTypes,
+      allowPatUnquotes: Boolean = this.allowPatUnquotes,
+      allowSpliceUnderscores: Boolean = this.allowSpliceUnderscores,
+      allowTermUnquotes: Boolean = this.allowTermUnquotes,
+      allowToplevelTerms: Boolean = this.allowToplevelTerms,
+      allowTrailingCommas: Boolean = this.allowTrailingCommas,
+      allowTraitParameters: Boolean = this.allowTraitParameters,
+      allowTypeLambdas: Boolean = this.allowTypeLambdas,
+      allowViewBounds: Boolean = this.allowViewBounds,
+      allowWithTypes: Boolean = this.allowWithTypes,
+      allowXmlLiterals: Boolean = this.allowXmlLiterals,
+      toplevelSeparator: String = this.toplevelSeparator,
+      allowNumericLiteralUnderscoreSeparators: Boolean =
+        this.allowNumericLiteralUnderscoreSeparators
+  ): Dialect = {
+    new Dialect(
+      allowAndTypes,
+      allowAtForExtractorVarargs,
+      allowCaseClassWithoutParameterList,
+      allowColonForExtractorVarargs,
+      allowEnums,
+      allowImplicitByNameParameters,
+      allowImplicitFunctionTypes,
+      allowInlineIdents,
+      allowInlineMods,
+      allowLiteralTypes,
+      allowMethodTypes,
+      allowMultilinePrograms,
+      allowOrTypes,
+      allowPatUnquotes,
+      allowSpliceUnderscores,
+      allowTermUnquotes,
+      allowToplevelTerms,
+      allowTrailingCommas,
+      allowTraitParameters,
+      allowTypeLambdas,
+      allowViewBounds,
+      allowWithTypes,
+      allowXmlLiterals,
+      toplevelSeparator,
+      allowNumericLiteralUnderscoreSeparators
+    )
+  }
+
+  // NOTE(olafur): Do not edit below here, these methods can remain unchanged.
+
+  override def productPrefix: String = "Dialect"
+  def productArity: Int = 0
+  def productElement(n: Int): Any = throw new IndexOutOfBoundsException(n.toString())
+
+  // Dialects have reference equality semantics,
+  // because sometimes dialects representing distinct Scala versions
+  // can be structurally equal to each other.
+  override def canEqual(that: Any): Boolean = this eq that.asInstanceOf[AnyRef]
+  override def equals(other: Any): Boolean = this eq other.asInstanceOf[AnyRef]
+  override def hashCode: Int = System.identityHashCode(this)
+  // Smart prettyprinting that knows about standard dialects.
+  override def toString = {
+    Dialect.standards.find(_._2 == this) match {
+      case Some((name, _)) => name
+      case None => s"Dialect(/* with custom parameters */)"
```

Comment:
`productIterator` using reflection? :)

Reply:
I made `productIterator` return empty list to avoid redundant boilerplate. `Dialect` should not be a `Product`, I added `@deprecated()` annotations on the `def product*` methods.

I changed this `toString` to return `Dialect()` instead. I never like the old `Dialect.toString`, it was very verbose for not benefit because it's hard to make sense of output like `Dialect(true, false, true, true, true, ....)`